relay-sampling/src/lib.rs

```diff
@@ -1,4 +1,52 @@
-//! Functionality for calculating if a trace should be processed or dropped.
+//! Sampling logic for performing sampling decisions of incoming events.
+//!
+//! Relay performs sampling of events in order to allow Sentry to offer performance at scale. The
+//! process of sampling ultimately comes down to generating a random number and given a sample rate
+//! checking whether the generated number is bigger or smaller than the sample rate.
+//!
+//! In order to determine the sample rate, Relay uses a [`SamplingConfig`] which contains a set of
+//! [`SamplingRule`]s that are matched against the incoming [`Event`] or [`DynamicSamplingContext`].
+//!
+//! # Trace and Transaction Sampling
+//!
+//! Relay samples both transactions looking at [`Event`] and traces looking at [`DynamicSamplingContext`]:
+//! - **Trace sampling**: ensures that either all transactions of a trace are sampled or none.
+//! - **Transaction sampling**: does not guarantee complete traces and instead applies to individual transactions.
+//!
+//! # Components
+//!
+//! The sampling system implemented in Relay is composed of several important components:
+//! - [`DynamicSamplingContext`]: a struct that contains the trace information.
+//! - [`FieldValueProvider`]: an abstraction implemented by [`Event`] and [`DynamicSamplingContext`] to
+//! expose fields that are read during matching.
+//! - [`SamplingRule`]: a rule that is matched against [`Event`] or [`DynamicSamplingContext`] that
+//! can contain a [`RuleCondition`] for expressing predicates on the incoming payload.
+//! - [`SamplingMatch`]: the result of the matching of one or more [`SamplingRule`].
+//!
+//! # How It Works
+//! - The incoming [`Event`] and optionally [`DynamicSamplingContext`] are received by Relay.
+//! - Relay fetches the [`SamplingConfig`] of the project to which the [`Event`] belongs and (if exists) the
+//! [`SamplingConfig`] of the root project of the trace.
+//! - The [`SamplingConfig`]s are merged together and the matching algorithm in
+//! [`SamplingMatch::match_against_rules`] is executed.
+//! - The sampling algorithm will go over each [`SamplingRule`] and compute either a factor or
+//! sample rate based on the [`SamplingValue`] of the rule.
+//! - The [`SamplingMatch`] is finally returned containing the final `sample_rate` and some additional
+//! data that will be used in `relay_server` to perform the sampling decision.
```

Comment:
I would add somewhere that the seed of the sampling decision is deterministic and depends on the trace ID / event ID, and that this allows us to

1. be consistent across a chain of relays (i.e. do not drop an event that was retained by a previous relay or vice versa),
2. be consistent across transactions of the same trace.